Feature/mw/fields2cover2

aosmw opened the pull request:
Update Fields2Cover to v2 api.  

Uses https://github.com/Fields2Cover/ortools_vendor/pull/2 and https://github.com/Fields2Cover/Fields2Cover/pull/126

This builds on behavior tree update [PR 43](https://github.com/open-navigation/opennav_coverage/pull/43) by @tonynajar.

tonynajjar committed:
Fix BT.CPP upgrade

tonynajjar committed:
Bug fixes

SteveMacenski committed:
Update opennav_coverage/src/visualizer.cpp

tonynajjar committed:
add readme

tonynajjar committed:
fix flake8

tonynajjar committed:
building fixes

tonynajjar committed:
testing with rolling

tonynajjar committed:
Revert "testing with rolling"

This reverts commit 3669e7d84d3065b0a149a20babe35fe5eba008b8.

tonynajjar committed:
test with rolling

tonynajjar committed:
Merge remote-tracking branch 'origin/main'

tonynajjar committed:
Add BTCPP_format="4" to trees

tonynajjar committed:
Use tree_->haltTree();

tonynajjar committed:
Revert "Use tree_->haltTree();"

This reverts commit aa2cf204ebceb08251c2cc26700bba21e78a3758.

tonynajjar committed:
Revert "Revert "Use tree_->haltTree();""

This reverts commit 668abc4318404d3d0ccce9a0aa92e08b3f089aaf.

DeccanLin committed:
Add gitignore; Add rviz config; (#51)

aosmw committed:
Update to fields2Cover v2.0.0 api

Signed-off-by: Mike Wake <michael.wake@aosgrp.com.au>

aosmw committed:
Add dependency on ortools_vendor

aosmw committed:
Port opennav_row_coverage to fields2cover v2

Signed-off-by: Mike Wake <michael.wake@aosgrp.com.au>

aosmw committed:
opennav_coverage: workaround non ament packaged Fields2Cover

Use tinyxml2_vendor properly.  Like in ros/pluginlib

IOW first find_package(tinyxml2_vendor REQUIRED) which sets up a module path
that provides a module for find_package(TinyXML2 REQUIRED) to work

Signed-off-by: Mike Wake <michael.wake@aosgrp.com.au>

aosmw committed:
opennav_coverage - fix tests for fields2cover v2

* fields2cover now initialises memory and provides getters/setters
so previous incorrectly/partially initialised data no longer throws

* Order of args for f2c::Random::generateRandField() changed to
  (area, sides), was (sides, area).  Tests now don't spin forever

Signed-off-by: Mike Wake <michael.wake@aosgrp.com.au>

SteveMacenski commented on .github/deps.repos:
```diff
@@ -3,3 +3,11 @@ repositories:
+    type: git
+    url: https://github.com/Fields2Cover/ortools_vendor.git
+    version: main
+  navigation2:
```
Lets not add in nav2, it causes the build to fail before we hit this work so I can't know if what you made compiles in an independent location. Plus this is a huge amount of code to compile

aosmw replied:
Yeah just circling back to this.  The need to build ortools is a drag.  Takes ages.  

I just tweaked an alternative inside Fields2Cover, by @Gonzalo-Mier, that grabs a release build of ortools. 

I wonder if that mechanism should be moved into ortools_vendor.  Do you know if it is necessary for ros packages to be built from source, or is this usage of a tarball acceptable?

aosmw replied:
I have done some more work on the Fields2Cover build.  

https://github.com/Fields2Cover/Fields2Cover/pull/133
https://github.com/Fields2Cover/Fields2Cover/pull/133/commits/f9f0a105d14bd16c850272c73c2cd735a3f75ca9

SteveMacenski commented on opennav_coverage/include/opennav_coverage/visualizer.hpp:
```diff
@@ -46,16 +46,16 @@ class Visualizer
     nav_plan_pub_ = rclcpp::create_publisher<nav_msgs::msg::Path>(
       node->get_node_topics_interface(),
-      "coverage_server/coverage_plan", rclcpp::QoS(1));
+      "coverage_server/coverage_plan", rclcpp::QoS(1).transient_local());
```
Update rviz config to match

aosmw replied:
I don't know about this one.  There is no rviz config it this repository that I can see.  It was a change by @tonynajjar

SteveMacenski replied:
Huh. I could have sworn someone added that recently 

Edit: oh, it was only for humble :roll_eyes:  https://github.com/open-navigation/opennav_coverage/commit/73fc308920bb4a71c23af39cb12086a528da047a

aosmw replied:
I cherry-picked the commit and amended the coverage_server/coverage_plan Durability Policy to Transient Local

https://github.com/open-navigation/opennav_coverage/pull/53/commits/9a38dc4ab8c7ad2a851f8ff04b2fa0c13974bab2

SteveMacenski commented on opennav_coverage/src/visualizer.cpp:
```diff
@@ -46,6 +46,7 @@ void Visualizer::visualize(
     for (unsigned int i = 0; i != utm_path->poses.size(); i++) {
       utm_path->poses[i].pose.position.x += ref_pt.getX();
       utm_path->poses[i].pose.position.y += ref_pt.getY();
+      utm_path->poses[i].header.frame_id = GLOBAL_FRAME;  // Necessary for Foxglove
```
No worries on the addition, but remove the comment

aosmw replied:
Done

SteveMacenski commented on opennav_row_coverage/test/test_server.cpp:
```diff
@@ -132,7 +132,7 @@ TEST(ServerTest, testDynamicParams)
 
   auto results = rec_param->set_parameters_atomically(
-    {rclcpp::Parameter("default_turn_point_distance", 0.25),
+    {rclcpp::Parameter("default_turn_point_distance", 0.25),  // Discretization ??
```
?

aosmw replied:
The turn_point_distance used to be a member variable of path_generator.  The API change appears to have changed this to a member function SetDiscretization on the class f2c::pp::TurningBase.

I can remove the comment.  The comment was a reminder to check if my interpretation of the api change was correct and whether the parameter name should be changed to better reflect what it is used for within Fields2Cover.  

https://github.com/Fields2Cover/Fields2Cover/blob/4b57a70dcac7f10cdd4760edbacdb0b48fdbc1c8/include/fields2cover/path_planning/turning_base.h#L69

aosmw replied:
removed comment.

SteveMacenski reviewed (changes requested):
Generally looks good to me, but you have some linting issues that CI is unhappy about that we need resolved and the build doesn't turn over :( 

I have no meaningful comments on your port to v2's API - that all looks great.

aosmw committed:
Revert addition of ortools_vendor to deps.repos

aosmw committed:
Remove unnecessary comment

aosmw committed:
Revert addition of navigation2 to deps.repos

aosmw committed:
Revert addition of ortools_vendor

aosmw committed:
uncrustify

aosmw committed:
Remove default_turn_point_distance == Discretization comment

aosmw commented:
Either the lines are too long for uncrustify or have wrong style for the cpplint.    

```
Error: /__w/opennav_coverage/opennav_coverage/opennav_coverage/include/opennav_coverage/utils.hpp:160:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
Error: /__w/opennav_coverage/opennav_coverage/opennav_coverage/include/opennav_coverage/utils.hpp:166:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
Error: /__w/opennav_coverage/opennav_coverage/opennav_coverage/include/opennav_coverage/utils.hpp:180:  If an else has a brace on one side, it should have it on both  [readability/braces] [5]
```

SteveMacenski commented:
> Either the lines are too long for uncrustify or have wrong style for the cpplint.

These are lines that this PR touches, please heed the formatting requirements :-)

aosmw committed:
cpplint/uncrustify style line lenght if/else style

Use intermediate variables for PathSectionType
to keep within line lenght limit requirements.

aosmw committed:
demo rviz coverage_plan/coverage_plan Transient Local

SteveMacenski commented:
CI is still failing, unfortunately, very early in this process - please look into that!

aosmw committed:
Put navigation2 back into .github/deps.repos for nav_msgs

This reverts commit 9c1ccfb57f0fd02506b2c86fd6439ae2268a5690.

aosmw committed:
DO NOT MERGE - temporarily pointing at aosmw fork of Fields2Cover

aosmw committed:
ci - attempt workaround rosdep not providing geometry_msgs

aosmw committed:
ci - attempt workaround rosdep not providing rosidl_default_generators

aosmw committed:
ci = guess a better base image that already has rclcpp in it

aosmw committed:
ci - install geographic_msgs

aosmw committed:
Fix spelling of geographic_msgs package

aosmw committed:
ci - install bondcpp and angles

aosmw committed:
Use ros-tooling/action-ros-ci@v0.3.11, rosdep-check: true, add test_msgs

aosmw committed:
ros-tooling/action-ros-ci@v0.3.11 not found go back to v0.3

Error: Unable to resolve action 'ros-tooling/action-ros-ci@v0.3.11', unable to find version v0.3.11'

aosmw committed:
ci - install behaviortree_cpp

aosmw committed:
Install diagnostic-updater

aosmw committed:
Try to get this to work with act(Run github actions locally).

https://nektosact.com/introduction.html
Run github actions locally.

It doesn't work locally for me. If fails on

vcs import --force --recursive src/ < package.repo
Could not checkout ref 'a077eb600a3700d1053482b3582822fa1b6e48db':
fatal: reference is not a tree: a077eb600a3700d1053482b3582822fa1b6e48db

aosmw committed:
More ci dependencies for nav2 build

aosmw committed:
ci - fix laser_geometry package name

aosmw committed:
ci - nav2 contrained smoother wants libceres-dev

aosmw committed:
ci - try libunwind-dev to try to get libgoogle-glog-dev to be happy for libceres-dev

aosmw committed:
ci - need robot-localization

SteveMacenski commented on .github/workflows/test.yml:
```diff
@@ -12,22 +12,63 @@ jobs:
     - uses: actions/checkout@v2
     - name: Install Cyclone DDS
       run: sudo apt install ros-${{ matrix.ros_distro }}-rmw-cyclonedds-cpp -y
+    - name: Install geometry_msgs
```
This should _not_ be necessary to call out each individual dependency, this is what rosdep is for

aosmw replied:
> CI is still failing, unfortunately, very early in this process with basic dependency resolution - please look into that!
I understand that rosdep is the way these items *SHOULD* be installed but rosdep was/is still?? not working for rolling on 22.04/jammy which is why I went in this direction . I know it is ugly and wrong and could be done in one step but I was taking bites at it one error at a time, the action build time getting progressively longer, now at 1h12m.   This is how I actually made progress in my own environments when rosdep got broken.  I use ansible to install the packages required.

There are currently no packages for ros-rolling-nav2* so this opennav_coverage with behavior tree v4 needs to build it.    Iron uses behaviourtree_cpp_v3 https://github.com/ros-navigation/navigation2/blob/4e16d3c7e3cfac08c4b79841f19570c40ebbe14b/nav2_behavior_tree/package.xml#L20 so not an option.

The rosdep workaround ["Attempt to fix Rolling 22.04 CI Until 24.04 Comes of Age"](https://github.com/ros-navigation/navigation2/commit/80bb5bff1488c0677efcc4254b7a89908c853ba0) that ros-planning/navigation2 uses was not immediately obvious on how or if it even should be applied to the opennav_coverage repository.  I assume duplicating that infra is not what you want, especially as its likely that the contents of this repository will be moved into nav2, soon after fields2cover packages correctly.

Is there is a better/faster solution where opennav_coverage can use a ci docker image of a built/installed rolling nav2 as a starting point?

In attempting to debug all this I was not able to get a local github actions environment setup. Do you use that? 
```
https://nektosact.com/introduction.html
It doesn't work locally for me. If fails on
vcs import --force --recursive src/ < package.repo
Could not checkout ref 'a077eb600a3700d1053482b3582822fa1b6e48db':
fatal: reference is not a tree: a077eb600a3700d1053482b3582822fa1b6e48db
```

NOTE: Before this is merged I also need to revert pointing this at my fork of fields2cover.  I am STILL trying to help fight the build farm to get packages of fields2cover to build  but we all live in different time zones and have other things going on so there is a lag between fix/apply/potentially revert that messes up this PR sometimes. See https://github.com/Fields2Cover/Fields2Cover/issues/136

I will look into how to tweak the action to only run the opennav_coverage tests and not the nav2 and fields2cover tests.

aosmw replied:
Related - Action ignores failed rosdep install https://github.com/ros-tooling/action-ros-ci/issues/864

aosmw replied:
Related - Rolling CI fails  https://github.com/ros-tooling/action-ros-ci/issues/844

aosmw replied:
Am I correct, albeit slow to realise it if so, rolling is no longer going to be supported/function on ubuntu 22.04?  Or is it just in a state of flux that has not yet been worked out?

SteveMacenski replied:
Did you try for Rolling on 24.04? That is what Rolling targets now, so it doesn't surprise me if you run into issues if you're not using a statically set rosdistro version pre-rolling update to 24.04. Instead of Jammy, use the Noble ones I think should fix that and not require specifically calling out every dependency this way. I think that's what the threads point to as well that you linked above! :smile: 

I haven't used the local Actions yet, no :(

SteveMacenski reviewed (changes requested):
It seems this is also running the full suite of tests on nav2/fields2cover etc. We should run tests only on the packages in this repository 

But the fact that this is building and running tests now is a great step! These 2 details (tests; rosdep) and we can merge this. Thanks for your time on this!

aosmw committed:
Build and test only what is necessary for this repo

aosmw commented:
Build/Test on rostooling/setup-ros-docker:ubuntu-jammy-ros-rolling-ros-base-latest
 2024-05-06T06:15:52.1236788Z E   TypeError: import_path() missing 1 required keyword-only argument: 'consider_namespace_packages'

A pytest incompatibilty

 https://github.com/ros2/launch/pull/766/files

aosmw committed:
Attempt build on noble, remove rosdep-check: true which doesn't perform install

aosmw committed:
Add separate jammy rolling action

aosmw committed:
Try removing python3-nose

mikolodz commented:
@aosmw Hello, thank you for your efforts! I was very excited to see your PR. What worries me is compatibility with 22.04 and with Humble. If it's going to work only with Jazzy & 24.04, it will be a major issue for many people, as 3rd party package compatibility is still far behind. People, especially working in SLAM area, are still using ROS1, so making use of a whole ROS project becomes gradually more complex or even impossible. It's even hard to understand where is the ROS project currently with the rolling development.

By any chance will this v2 API update be compatible with Humble or should I start preparing for a big upgrade? I tried to build your code along with my Humble project, but I'm facing a lot of issues. I fixed a dozen of them and now I'm stuck again after building it successfully (I had to use nav2 humble source), on linking. Did you try building it on Humble or do you plan to make it working on it?

aosmw committed:
Try running apt-get update to get latest python3-rosdep

image is currently 7 weeks old

aosmw committed:
pip uninstall nose

colcon finds nosetests which uses imp which is not supported in python3.12

aosmw committed:
Fix apt update, use PIP_BREAK_SYSTEM_PACKAGES=1 for pip command

aosmw committed:
fix PIP_BREAK_SYSTEM_PACKAGES env var

aosmw commented:
Hi @mikolodz, 

My environment is humble on ubuntu22.04.  However I do build personal snapshots of navigation2, fields2cover and opennav_coverage and robot_localisation that I periodically bring close to rolling since I really want to use and understand how to tune nav2_mppi_controller, nav2_smac_planner and nav2_behavior_tree.  I do worry about the improvements/fixes to the lower layers rclcpp/rcl that I may be missing out on.

This PR and work on Fields2Cover packaging is my attempt at giving back.  I am dipping in and out of this PR fighting with the flux of noble/ubuntu 24.04/python3.12 removals(imp)/rosdep breakage/colon empyv4 incompatibility.  Alot of my recent changes to this PR are fighting the github action build, most of which should ultimately be squashed away and probably removed once the lower layers of infrastructure are again functional.  

The problems on the bleeding edge are deep but I am seeking to contribute fixes/ideas/workarounds.  I thought initially I might get away with a tweak or two to.  I am being drawn in and going deeper down into the stack understanding more and more how the infrastructure works.  A good thing.

BTW I also would like to use some of the new awesome SLAM stuff like LIO-EKF (https://github.com/YibinWu/LIO-EKF)  (Currently ROS1)   

As for compatible with Humble.  The policy appears to be, get it working on rolling first, and then potentially backport.

aosmw committed:
more worflow

aosmw committed:
workflow

aosmw commented:
noble_rolling now hitting warning as error in behaviortree_cpp.

 /opt/ros/rolling/include/behaviortree_cpp/controls/switch_node.h:91:15: error: declaration of ‘ports’ shadows a previous local [-Werror=shadow]
     91 |     PortsList ports;

This is fixed in https://github.com/BehaviorTree/BehaviorTree.CPP/commit/97c3f6ab5982ccfd487911be15c29e3ac84acbac 

Which is in https://github.com/BehaviorTree/BehaviorTree.CPP/releases/tag/4.6.0

mikolodz commented:
Thank you @aosmw

I truly admire your patience. I've been working with SC-LIO-SAM (ROS1) for a while and got it running with gazebo & Nav2. Wasn't as easy as I expected :) Currently, I'm stuck with the fields2cover, and this is why I'm here asking you questions and observing your efforts.

Do you think I could use your snapshots to build your fork for Humble? As I said before, I have some linking issues and felt like I'm not sure if I should invest another day or two getting through it not knowing if it will work at all. But now, when you confirm that you're using a similar setup as I use, it gives me new motivation to give it a try.

aosmw committed:
Attempt to manipulate rosdep index for jammy back to Feb 28,2024

aosmw committed:
Try single quotes for sed

aosmw committed:
action syntax

aosmw committed:
action syntax sed

aosmw committed:
Run through yamllint

aosmw committed:
Adjust noble_rolling to use pre-release binaries

This is so that nav2 can build with -Werror against
behaviortree_cpp v4.6

aosmw committed:
noble_rolling needs libyaml-cpp-dev rosdep problem?

aosmw committed:
jammy_rolling: workaround libgoogle-glog-dev libunwind-dev libceres-dev

aosmw commented:
noble_rolling is not finding libyaml-cpp-dev via  https://github.com/ros2/yaml_cpp_vendor.   
 
 [ 22%] Linking CXX executable test_map_saver_publisher
  /usr/bin/ld: ../../libmap_io.so: undefined reference to `YAML::Emitter::StartedScalar()'
  /usr/bin/ld: ../../libmap_io.so: undefined reference to `vtable for YAML::BadConversion'
  /usr/bin/ld: ../../libmap_io.so: undefined reference to `YAML::detail::node_data::insert_map_pair(YAML::detail::node&, YAML::detail::node&)'
  /usr/bin/ld: ../../libmap_io.so: undefined reference to `YAML::ostream_wrapper::write(std::__cxx11::basic_strin

jammy_rolling  - back to same pytest incompatability - https://github.com/ros2/launch/pull/766/files
/opt/ros/rolling/lib/python3.10/site-packages/launch_testing/pytest/hooks.py:178: in find_launch_test_entrypoint
      module = import_path(path, root=None)
  E   TypeError: import_path() missing 1 required keyword-only argument: 'consider_namespace_packages'

SteveMacenski commented:
> What worries me is compatibility with 22.04 and with Humble.

There is a separate humble branch - so that all should be more or less fine! 

> are still using ROS1, so making use of a whole ROS project becomes gradually more complex or even impossible.

There is no intention for support for ROS 1 

> By any chance will this v2 API update be compatible with Humble or should I start preparing for a big upgrade?

For maintenance reasons, the humble branch will stay on V1 since that was what it was released under and v2 has breaking changes. With that said, I'd be happy to have a `humble-v2` branch that folks on Humble can use with v2 if someone opens a PR once this is all wrapped up. I just don't want to push API breaking changes to _everyone_ except those that want that in exchange for the breakage. 

> This PR and work on Fields2Cover packaging is my attempt at giving back.

And I genuinely thank you for this @aosmw! I spend alot of my time fighting the good fight working on things like this, so the help is really, truly appreciated! 

> This is fixed in https://github.com/BehaviorTree/BehaviorTree.CPP/commit/97c3f6ab5982ccfd487911be15c29e3ac84acbac

Ugh. Build from source until that is released too.... I guess...

> Needs something like this https://github.com/ros-navigation/navigation2/pull/4267

Precisely!

SteveMacenski commented:
Any word?

aosmw commented:
I was going to wake this up again now that https://github.com/ros-navigation/navigation2/pull/4298 has landed.
Most of what is needed is just stripping out the workarounds in .github

I was also going to look into how to point the github build at the ghcr.io/ros-navigation/navigation2:main docker image.

aosmw committed:
Abandon rolling on jammy

aosmw committed:
Strip away build on influx rolling workarounds

aosmw committed:
Comment out fields2cover from deps.repos, attempt to build against packages

aosmw committed:
Add ortools_vendor to deps.repos

aosmw committed:
gh action: install nodejs for local build nektos/act

nodejs in docker image is required to enable enable
local debugging of github actions with
https://github.com/nektos/act

aosmw committed:
Build fields2cover from source again

This is to try to eliminate the possibility that
we are fetching a bad package.

aosmw committed:
nav2_constrained_smoother has conflicted libunwind-dev, libgoogle-glog-dev for libceres-dev

aosmw commented:
The build against the source of fields2cover works, ie using the .github/deps.repos mechanism.   
The build against the packaged version of fields2cover for rolling on 24.04 currently does not.

HOWEVER, 
There remains a package conflict to negotiate for nav2_constrained_smoother and nav2_smac_planner that is not resolved by rosdep.  libceres-dev wants libgoogle-glog-dev that wants libunwind-dev but the docker image has libunwind7-dev.

ALSO,
The packaged install location of the fields2cover dependent libraries is pretty ugly and fails to find its libabsl libraries properly.  Looking for them in /opt/ros/lib/lib rather than in /opt/ros/lib/x86_64-linux-gnu where they currently reside.  

IOW,
The current install of ros-rolling-fields2cover 2.0.0-13noble.20240514.144256 put the ortools/absl/protobuf/fields2cover libraries in /opt/rolling/lib/x86_64-linux-gnu rather than a more conventional/contained ros location.  They should be properly vendored into /opt/ros/rolling/opt/ortools

FYI,
I now have a way to poke around in the github action build by running it using https://github.com/nektos/act.  It requires that the image being used has nodejs installed.  Once it is running I find its container name using docker ps, and then get a bash shell inside it using `docker exec -it $(docker ps --latest --quiet) bash`

CONCLUSION,
Close but still so far.

SteveMacenski commented:
>  The build against the packaged version of fields2cover for rolling on 24.04 currently does not.

Great... I thought we'd mentioned that to fix it, but I suppose still no dice. 

>  There remains a package conflict to negotiate for nav2_constrained_smoother and nav2_smac_planner that is not resolved by rosdep. libceres-dev wants libgoogle-glog-dev that wants libunwind-dev but the docker image has libunwind7-dev.

I'm a little confused by that considering this builds fine in Nav2's CI and for me locally with a 24.04 docker image

SteveMacenski commented:
Perhaps it would be easier to manually setup the build [like here](https://github.com/rsasaki0109/lidar_localization_ros2/pull/38/files) rather than using the ROS tooling wrappers?

aosmw commented:
The image used in the github workflow is currently `rostooling/setup-ros-docker:ubuntu-noble-ros-rolling-ros-base-latest`
That is the likely cause of the libceres-dev, libgoogle-glog-dev, libunwind-dev, libunwind7-dev dance at the moment.   

The main problem of using the Fields2Cover package, (2.0.0-13noble.20240514.144256)  right now is the command line flags that the ros buildfarm uses and what they do to the locations of where things get installed prior to creating the debian package.  

Fields2Cover does not use ament_cmake and I when I started trying to help, I tried to workaround what was there.  I introduced the use of cmake `include(GNUInstallDirs) and used the variable CMAKE_INSTALL_LIBDIR which is overriden by the ros packaging command in a way that is not playing nicely.

Digging down into it the command that is getting run by the ros build farm to create the package  - https://build.ros2.org/job/Rbin_uN64__fields2cover__ubuntu_noble_amd64__binary/77/console and wrapping it in a script to reproduce locally 

```
#!/bin/bash

mkdir -p build
cd build
DEB_PYTHON_INSTALL_LAYOUT=deb cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DFETCHCONTENT_FULLY_DISCONNECTED=ON "-GUnix Makefiles" -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu -DBUILD_TESTING:BOOL=OFF -DCMAKE_INSTALL_PREFIX=/optmw/ros/rolling -DCMAKE_PREFIX_PATH=/optmw/ros/rolling -DBUILD_TESTING=OFF -DFETCHCONTENT_FULLY_DISCONNECTED=OFF ..

make -j8
sudo make install
```

Notice the `-DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu`
Notice that i tweak the install prefix to /optmw  isolate it on my local dev system.

I have some ideas to get it to play nicely (in a ros package install) that I am playing around with.

My aim is to find a way to get the ortools sfuff installed in /opt/ros/rolling/opt/f2c_ortools while still working for a non-ros/non-ament build.

THIS https://github.com/Fields2Cover/Fields2Cover/issues/146 (remove libabsl-dev from Fields2Cover/package.xml) might be a simple fix for now but the way the Fields2Cover package is dumping ortools artifacts into /opt/ros/rolling should really be improved so I will keep plugging away at it.

SteveMacenski commented:
> I have some ideas to get it to play nicely (in a ros package install) that I am playing around with.

Great!! For the meantime though would it be useful to just build F2C in CI here from a `deps.repos` file that we use tagged at a version we support for a given distribution? 

But fixing F2C to have the binaries work well is useful. If we're having issues here, I presume anyone else trying to use the binaries would also have a similar issue.

aosmw committed:
Ignore vim .swp files

aosmw committed:
deps.repos lock to tags for ortools_vendor and Fields2Cover

NOTE: Fields2Cover still pointed at fork.

MarcM0 commented:
Sorry to bother but, from reading the history, this works on older versions of ROS and you are currently just fighting the autotester?

aosmw commented:
@MarcM0 This PR targets rolling, which was in a fair bit of flux when I was last touching this.   Its settled now so worth a revisit.  I won't be looking at it for a while though.  Next steps would be to update it to the latest rolling and check the state of the fields2cover packages.

Jakubach commented:
@aosmw I managed to build f2c with that and colcon build the `ortools_vendor` but while buidling `opennav_coverage` I get this error. I am using your forks packages and doing it on ROS2 humble. Do you know how could it be solved? 

```
Starting >>> opennav_coverage
[Processing: opennav_coverage]                              
--- stderr: opennav_coverage                                
CMake Error at /usr/share/cmake-3.22/Modules/CMakeFindDependencyMacro.cmake:47 (find_package):
  Could not find a package configuration file provided by "ortools" with any
  of the following names:

    ortoolsConfig.cmake
    ortools-config.cmake

  Add the installation prefix of "ortools" to CMAKE_PREFIX_PATH or set
  "ortools_DIR" to a directory containing one of the above files.  If
  "ortools" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  /home/Jakub/ros2_ws/install/fields2cover/lib/cmake/fields2cover/Fields2CoverConfig.cmake:18 (find_dependency)
  CMakeLists.txt:19 (find_package)


---
Failed   <<< opennav_coverage [58.9s, exited with code 1]

Summary: 1 package finished [7min 38s]
  1 package failed: opennav_coverage
  2 packages had stderr output: fields2cover opennav_coverage
```

mikolodz commented:
> @MarcM0 This PR targets rolling, which was in a fair bit of flux when I was last touching this. Its settled now so worth a revisit. I won't be looking at it for a while though. Next steps would be to update it to the latest rolling and check the state of the fields2cover packages.

Hello @aosmw, how are you doing? Do you think you will be able to give it a try in the near future?

lianzhijun17 commented:
@SteveMacenski @tonynajjar @aosmw @mikolodz
Hello,
Thank you very much for your time and assistance!
I am currently using the Humble branch. When I use the F2C v1.2.1 API, the simulation runs perfectly in Gazebo. However, when I switch to the F2C v2.0.0 API, I encounter an issue during the simulation. Although I can see the coverage path and the robot moves to the start point, it stops there and does not proceed further until the progress checker times out. The /cmd_vel topic shows a very low speed of 0.0005, which seems to be the root of the problem.
I am unsure how to resolve this issue and would greatly appreciate any guidance or suggestions you might have.
Thank you again for your help!
![stop_at_start_point](https://github.com/user-attachments/assets/ef4c5978-5517-492d-aec9-ebdbdc9d1461)
![stop_at_start_point_rviz](https://github.com/user-attachments/assets/bfe91f6b-20bd-457d-89d9-a1a172ceddfb)
![image](https://github.com/user-attachments/assets/52e90799-b426-458a-b495-84af32f11813)

u33549 commented:
Offering support for Fields2Cover v2 API Integration / Reviving PR #53

Hello @SteveMacenski, @aosmw, and team, 
I have been reviewing this PR and the history of migrating opennav_coverage to the Fields2Cover v2 API. I see that due to packaging bottlenecks and binary/dependency conflicts on Noble/Rolling at the time, the project recently merged support for F2C v1.2.1 on Jazzy/Rolling via #82.

I am very interested in contributing to this repository and helping revive the transition to F2C v2.0.0+.

Before diving into the code, I would like to ask:

Is there an ongoing active branch or a new roadmap for the v2 migration, or should we start fresh from the current main branch state?

Has the F2C Debian packaging issue (the libabsl / GNUInstallDirs path mismatch on the ROS buildfarm) been resolved upstream in Fields2Cover, or do we still need to vendor it via deps.repos for CI?

Regarding the runtime issue mentioned by @lianzhijun17 (where the robot stalls at the start point with near-zero /cmd_vel), has anyone pinpointed whether this is a local Nav2 controller configuration issue or an architectural change in how F2C v2 returns path coordinates?

I have a local setup with Ubuntu 22.04/24.04 and ROS 2, and I'd be glad to take over the heavy lifting, fix the linting/braces issues, and debug the simulation tracking. Looking forward to your guidance!

SteveMacenski commented:
> Is there an ongoing active branch or a new roadmap for the v2 migration, or should we start fresh from the current main branch state?

There is not current active branch, so if you wanted to start clean or start with something related to this, either is OK by me! I think its up to you and what you think is best / most efficient :-) 

> Has the F2C Debian packaging issue (the libabsl / GNUInstallDirs path mismatch on the ROS buildfarm) been resolved upstream in Fields2Cover, or do we still need to vendor it via deps.repos for CI?

I think we should plan to build from source to pin to a release version, but you can/should test. 

Happy to have the help!

u33549 commented:
Thanks for the quick and encouraging response, @SteveMacenski! 🙏

Since there's no active branch and it's open to whatever is most efficient, here's how I plan to approach it:
Starting point: humble / Ubuntu 22.04. I'm most comfortable in that environment and prefer working on 22.04, so I'd like to land the v2 migration there first, and only look at Jazzy/Rolling after that. 
I've audited the entire F2C surface in the repo (only opennav_coverage and opennav_row_coverage actually touch F2C — 9 production sources/headers + 10 test files), and I've split the work into two clearly separated tracks:

1. UPDATE — mandatory v2 API migration (no behavior change, just compile + run on v2)
These are the existing modules re-wired to the v2 API:

types.hpp — clean up the duplicate LineString typedef; re-validate all F2C* aliases against v2 headers.
robot_params.hpp (+ the two dynamic-param callbacks in coverage_server.cpp and row_coverage_server.cpp) — public fields → setters: robot_width/op_width/linear_curv_change → setWidth()/setCovWidth()/setMaxDiffCurv(), setMinRadius() → setMinTurningRadius(), and handle the parameterized Robot constructor.
swath_generator.cpp — step_angle public field → setStepAngle().
path_generator.cpp — searchBestPath() → planPath(); turn_point_dist → TurningBase::setDiscretization().
headland_generator.cpp — verify generateHeadlands / ConstHL signatures.
route_generator.cpp — verify genSortedSwaths / setSpiralSize / setCustomOrder.
coverage_server.cpp — F2CField public fields → getters (.coord_sys → getCRS(), .field → getField()); verify Parser::importFieldGml / Transform::transformToUTM.
utils.hpp — path.states → getStates(); re-validate PathState/PathSectionType, F2CField/F2CCell constructors, Transform::transformToPrevCRS, moveTo.
visualizer.cpp — verify geometry primitives (getGeometry, Point + Point, startPoint/endPoint).
opennav_row_coverage/utils.hpp — the F2C↔GDAL/OGR bridge + CRS handling (most fragile integration point).
row_swath_generator.cpp — StartPoint()/EndPoint() → camelCase startPoint()/endPoint(); verify Swath(LineString, width, id) ctor and reversePoints().
Build — find_package(Fields2Cover 2.0 REQUIRED), Fields2Cover::Fields2Cover link target, package.xml rosdep key.
Tests — update the 10 affected test files to the new API.

2. UPGRADE — new v2 capabilities to add after the migration lands
These are genuinely new features F2C v2 unlocks that the current code doesn't use yet (ordered low → high effort):

NSwathModified objective — fast swath-count objective, speeds up the brute-force angle sweep.
Velocity profile + driving direction — expose PathState::velocity and dir (FORWARD/BACKWARD) to the ROS output.
Path post-processing — discretize/populate/reduce, plus getTaskTime() for an ETA in the action feedback/result.
HL_SWATH type + headland swaths — treat headland rings as drivable swaths.
F2CRoute + inter-swath connections — real transition paths between swaths via planPath(robot, route, curve).
Facade (planCovPath / Options) — an optional "single-call" simple-mode pipeline alongside the granular one.
Decomposition (TrapezoidalDecomp / BoustrophedonDecomp) — split non-convex (L/U-shaped) fields into convex sub-cells, drastically reducing swath count.
Multi-cell TSP routing (RoutePlannerBase) + start/end point pinning (setStartAndEndPoint) — OR-Tools-based routing through headlands across multiple cells, starting from the robot's current pose. This is the biggest architectural win.


My plan is to get Track 1 (Update) fully compiling, tested, and validated against the demo package first, then layer Track 2 (Upgrade) features incrementally so each can be reviewed on its own.
I'll start a clean branch off main/humble and open a draft PR early so you can track progress. Looking forward to it!

SteveMacenski commented:
https://github.com/open-navigation/opennav_coverage/pull/105#pullrequestreview-4583079031 to supersede
